engines/e_devcrypto.c

```diff
@@ -698,70 +716,97 @@ static int digest_init(EVP_MD_CTX *ctx)
     const struct digest_data_st *digest_d =
         get_digest_data(EVP_MD_CTX_type(ctx));
 
-    digest_ctx->init_called = 1;
 
     memset(&digest_ctx->sess, 0, sizeof(digest_ctx->sess));
     digest_ctx->sess.mac = digest_d->devcryptoid;
-    if (ioctl(cfd, CIOCGSESSION, &digest_ctx->sess) < 0) {
-        SYSerr(SYS_F_IOCTL, errno);
-        return 0;
-    }
     return 1;
 }
 
 static int digest_op(struct digest_ctx *ctx, const void *src, size_t srclen,
                      void *res, unsigned int flags)
 {
     struct crypt_op cryp;
+    int ret;
 
     memset(&cryp, 0, sizeof(cryp));
+    if (ctx->init_called != 1) {
+        if (ioctl(cfd, CIOCGSESSION, &ctx->sess) < 0) {
+            SYSerr(SYS_F_IOCTL, errno);
+            return 0;
+        }
+        ctx->init_called = 1;
+    }
     cryp.ses = ctx->sess.ses;
     cryp.len = srclen;
     cryp.src = (void *)src;
     cryp.dst = NULL;
     cryp.mac = res;
     cryp.flags = flags;
-    return ioctl(cfd, CIOCCRYPT, &cryp);
+    if ((ret = ioctl(cfd, CIOCCRYPT, &cryp)) < 0)
+        SYSerr(SYS_F_IOCTL, errno);
+    return ret;
 }
 
 static int digest_update(EVP_MD_CTX *ctx, const void *data, size_t count)
 {
     struct digest_ctx *digest_ctx =
         (struct digest_ctx *)EVP_MD_CTX_md_data(ctx);
+    void *new_data;
+    int ret = 0;
 
     if (count == 0)
         return 1;
 
     if (digest_ctx == NULL)
         return 0;
 
-    if (EVP_MD_CTX_test_flags(ctx, EVP_MD_CTX_FLAG_ONESHOT)) {
-        if (digest_op(digest_ctx, data, count, digest_ctx->digest_res, 0) >= 0)
-            return 1;
-    } else if (digest_op(digest_ctx, data, count, NULL, COP_FLAG_UPDATE) >= 0) {
-        return 1;
+    if (EVP_MD_CTX_test_flags(ctx, EVP_MD_CTX_FLAG_ONESHOT))
+        return (digest_op(digest_ctx, data, count, digest_ctx->digest_res, 0) >= 0);
```

Comment:
Shouldn't this use UPDATE | FINAL as flags?

Reply:
The original cryptodev implementation did not have update/final support.  It was added later, so the operation was initially done without the flags.  Documentation is sparse, so there's no guarantee, but looking at `main.c` in linux-cryptodev, it looks like it works with both flags as well.  I'm not sure it needs a reset, if so, a `COP_FLAG_RESET` would be needed.  When running without flags, it performs a reset as the first step, then it does update, and calls final.  I'd go with history and leave the flags zeroed.  Notice that this call was originally written (by me) without them, as was the case in the old cryptodev engine.